### PR TITLE
Artifact Metadata Driven Docs Release

### PR DESCRIPTION
**Pull request by scbedd**
This PR handles the "readme" side, but the script will get larger as we add additional capabilities here. We have all the necessary metadata from the artifact parse that we use for creating tags and git releases.

Essentially:
 * pull down the doc repo, if the branch we want exists grab it
 * Apply the readme update for each discovered artifact in the release
 * Push the readme updates to `smoke-test-rdme` or `smoke-test-preview-rdme`.
 * Submit PullRequest if one doesn't already exist.

[Example of a Run](https://github.com/MicrosoftDocs/azure-docs-sdk-node/pull/729/files)

There is an issue that I seemingly can't figure out that looks like the following when I re-run the script for an already existing + updated readme. The git diff looks empty, but gives the error

```
C:/Users/sbedd/AppData/Local/Temp/tmpD018.tmp:20: trailing whitespace.
ms.service:
C:/Users/sbedd/AppData/Local/Temp/tmpD018.tmp:23: trailing whitespace.
 # Azure Core HTTP client library for JavaScript - Version 1.1.1
error: patch failed: docs-ref-services/core-http-readme.md:1
error: docs-ref-services/core-http-readme.md: patch does not apply
```

Not certain what's going on with that, need to resolve before merge.

@maggiepint used your script basically directly for the readme mods. Thank you!

### CODE REVIEW

File: eng/common/scripts/git-branch-push.ps1 [hunk 1]
```diff
@@ -51,8 +51,19 @@ if ($LASTEXITCODE -ne 0)
     exit $LASTEXITCODE
 }
 
-Write-Host "git checkout -b $PRBranchName"
-git checkout -b $PRBranchName
+# 0 if found, false if not
+git show-ref --verify --quiet refs/heads/$PRBranchName
```
Comment (JimSuplizio) [inline eng/common/scripts/git-branch-push.ps1:55]: /CC @weshaggard 
Two things
1. Why is this change necessary?
2. I'm not sure you need the refs/heads/$PRBranchName or if the $PRBranchName is sufficient.

Comment (scbedd) [reply eng/common/scripts/git-branch-push.ps1:55]: I'm accumulating changes in the `PR` branch. My understanding of this script is that it'll always attempt to check the branch as new in the current iteration. Unfortunately, when you fetch upstream, it also picks up that PR branch. After that, the git checkout -b will fatal with `branch already exists`. 

I added the change so that when it DOES find the branch I'm attempting to push to, it won't just crash out.

Hmm. I may be able to accumulate in smoke-test only and git-branch-push to the other branch. I'm not sure. WIll try it when I'm back at my machine.

Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:55]: `git checkout -b $PRBranchName` here only create the branch in the local clone and should never fail on a fresh clone that hasn't already created that branch. If the branch already exists from upstream this command will track that upstream branch and make the branch pointers the same. So I don't think there is any need to check for this branch.

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+    $headers = @{
+      "Content-Type"  = "application/json"
+      "Authorization" = "token $($env:GH_TOKEN)"
+    }
+
+    FireAPIRequest -url $url -body $body -headers $headers -method "Post"
+  }
+}
+
+function FireAPIRequest($url, $method, $body = $null, $headers = $null) {
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:39]: To help folks when they read this I'd rename this function `Invoke-RestMethodWithRetries`. Actually Invoke-RestMethod (https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-restmethod?view=powershell-7) already supports a retry parameter that we should probably use here and then this method simply becomes a wrapper for handling exceptions. Perhaps `Invoke-RestMethodWithHandling`?

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+  foreach ($pkgInfo in $pkgList) {
+    Write-Host "Creating release $($pkgInfo.Tag)"
+
+    $releaseNotes = ""
+    if ($pkgInfo.ReleaseNotes[$pkgInfo.PackageVersion].ReleaseContent -ne $null) {
+      $releaseNotes = $pkgInfo.ReleaseNotes[$pkgInfo.PackageVersion].ReleaseContent
+    }
+
+    $isPrerelease = $False
+    if ($pkgInfo.PackageVersion -match $VERSION_REGEX) {
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:15]: Doesn't have to be part of this change but at some point we should switch to using our common versioning helpers at https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/SemVer.ps1

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    ReadmeContent  = $readmeContent
+  }
+}
+
+# Returns the maven (really sonatype) publish status of a package id and version.
+function IsMavenPackageVersionPublished($pkgId, $pkgVersion, $groupId) {
+  try {
+
+    $uri = "https://oss.sonatype.org/content/repositories/releases/$groupId/$pkgId/$pkgVersion/$pkgId-$pkgVersion.pom"
+    $pomContent = Invoke-RestMethod -Method "GET" -Uri $uri
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:113]: We should use our wrapper Invoke-RestMethod.

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+  }
+}
+
+# Returns the nuget publish status of a package id and version.
+function IsNugetPackageVersionPublished($pkgId, $pkgVersion) {
+
+  $nugetUri = "https://api.nuget.org/v3-flatcontainer/$($pkgId.ToLowerInvariant())/index.json"
+
+  try {
+    $nugetVersions = Invoke-RestMethod -Method "GET" -Uri $nugetUri
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:239]: Should we use our wrapper Invoke-RestMethod.

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+    Deployable     = $true
+    ReleaseNotes   = $releaseNotes
+    ReadmeContent  = $readmeContent
+  }
+}
+
+# Returns the pypi publish status of a package id and version.
+function IsPythonPackageVersionPublished($pkgId, $pkgVersion) {
+  try {
+    $existingVersion = (Invoke-RestMethod -Method "Get" -Uri "https://pypi.org/pypi/$pkgId/$pkgVersion/json").info.version
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:314]: Should we use our wrapper Invoke-RestMethod.

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+  }
+
+  return ($pathsWithComplexity | Sort-Object -Property Complexity)[0].Path
+}
+
+# Parse out package publishing information given a .tgz npm artifact
+function ParseNPMPackage($pkg, $workingDirectory) {
+  $workFolder = "$workingDirectory$($pkg.Basename)"
+  $origFolder = Get-Location
+  mkdir $workFolder
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:155]: As we start to try and reuse these scripts we should try and switch to use the cmdlets directly because alias' like these often times do different things on different OS's. Some OS's have a built in command others use the cmdlet and they generally have different logic and options, so we should try to always use the cmdlet if possible.

File: eng/common/scripts/git-branch-push.ps1 (hunk 1 above)
Comment (JimSuplizio) [reply eng/common/scripts/git-branch-push.ps1:55]: @weshaggard That's what I thought and why I'd asked the question. If that had been a problem we'd have seen it during the last few releases as there are a number of pipelines that create add commits to the same branch. I was more curious as to whether or not @scbedd was seeing something going awry that was unique to his scenario.

File: eng/common/scripts/release-docs-metadata.ps1 [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    }
+  }
+}
+
+function SyncDocRepo($docRepo, $workingDirectory, $sourceBranch, $prBranch) {
+  # get the doc repo, looking for the rdm branch if it exists
+  $docRepo = "https://github.com/$docRepo"
+  $repoRoot = (Join-Path $workingDirectory "repo/" )
+
+  _execute("git clone $docRepo $repoRoot")
```
Comment (weshaggard) [inline eng/common/scripts/release-docs-metadata.ps1:47]: This function looks to be doing more then we probably need it to do. I assume all you want to do is clone the repo and checkout a particular branch, if it exists otherwise create it.  Similar to my comments in my other file I don't expect that you need to actually check for the branch, you simply check it out and it will either track the upstream version if it exists and if not it will create a new branch from the clone point.

Comment (weshaggard) [reply eng/common/scripts/release-docs-metadata.ps1:47]: Might be worth having a look at https://github.com/Azure/azure-sdk-tools/blob/master/eng/pipelines/templates/steps/sync-directory.yml which is what we use to sync the eng/common directory.

File: eng/common/scripts/release-docs-metadata.ps1 [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    if ($adjustedContent) {
+      try {
+        Push-Location $repoLocation
+        Set-Content -Path $readmeLocation -Value $adjustedContent -Force
+
+        Write-Host "git add -A ."
+        git add -A .
+
+        # commit the changes to the PR Branch. If others have pushed to the target branch we want to rebase off it and commit our changes regardless
+        & $branchScriptLocation -PRBranchName $prBranch `
```
Comment (weshaggard) [inline eng/common/scripts/release-docs-metadata.ps1:189]: I'd suggest you only have this script do the source updates and move the clone, branching, and pull-request work out into the yml pipeline that will be doing this work.

File: eng/common/scripts/git-branch-push.ps1 (hunk 1 above)
Comment (scbedd) [reply eng/common/scripts/git-branch-push.ps1:55]: Ok. I'll try removing this piece and see what happens. I must be missing something because this absolutely failed on my box without the code change.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 3 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:15]: You're right. I wasn't attempting to introduce new bugs with this PR. 

That is why in the new script I absolutely use `SemVer.ps1`

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 7 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:155]: I don't understand what you mean by this comment. `Get-Location` IS the cmdlet versus `pwd` which would be the local alias right?

File: eng/common/scripts/release-docs-metadata.ps1 (hunk 9 above)
Comment (scbedd) [reply eng/common/scripts/release-docs-metadata.ps1:189]: Hmm. I don't really agree, but I'll try and see how it looks.

File: eng/common/scripts/release-docs-metadata.ps1 (hunk 8 above)
Comment (scbedd) [reply eng/common/scripts/release-docs-metadata.ps1:47]: I _don't_ want to only checkout a particular branch. I want to accumulate readme changes to a _specific_ branch. I don't particularly want to file a new PR every time. See where I'm coming from?

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 6 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:314]: Actually. Yes. I'll update it.

File: eng/common/scripts/git-branch-push.ps1 (hunk 1 above)
Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:55]: feel free to ping me and I can explain why I think it failed locally but won't fail on ci machines. The biggest reason is locally I suspect you are re-using the clone which already has the branch created, but in ci it is always a fresh clone.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 7 above)
Comment (weshaggard) [reply eng/common/scripts/artifact-metadata-parsing.ps1:155]: For this particular comment I was looking at `mkdir` which on windows is a alias to the cmdlet but on linux it uses the shell native linux command and they have differences. Namely one of them will create subdirectories and the other will not, which has bitten me in the past (see https://github.com/Azure/azure-sdk-tools/pull/357 for some examples).  To be clear this isn't something you need to change with this PR but as we refactor theses scripts an expect them to run on multiple OS's we will start to hit these issues.

File: eng/common/scripts/release-docs-metadata.ps1 (hunk 8 above)
Comment (weshaggard) [reply eng/common/scripts/release-docs-metadata.ps1:47]: You don't need a new PR for each one and you can totally create a commit per change if you like however if you are making multiple commits from the same clone you will definitely hit the branch creation problem you were seeing. 

Is there a reason you want multiple commits from the same set of readme updates? Won't they be squashed on merged? If we do feel that is an interesting scenario I think we can come up with some options. I'll think about that a scenario a little more.

File: eng/common/scripts/release-docs-metadata.ps1 (hunk 9 above)
Comment (weshaggard) [reply eng/common/scripts/release-docs-metadata.ps1:189]: It is generally a good idea to have the script be single purposed and if you can imagine yourself wanting to do this from the command  line without all the orchestration of cloning or creating pull requests then it seems like it would be a good idea to have them be separate steps.

File: eng/common/scripts/git-branch-push.ps1 (hunk 1 above)
Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:55]: Based on your other comments if you are trying to make multiple commits the the branch then I'd suggest you do all those commits in another temp branch before calling this script and then use this script to handle the pushing of the real branch for doing the pull request.

Comment (scbedd) [reply eng/common/scripts/git-branch-push.ps1:55]: Hmm. My goal here is to get zero-touch to the doc repo from each release. I just pushed a refactor that moves to a yml abstraction layer + does a readme per.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 2 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:39]: Yeah this is fair. Given that I'm not getting this in this release, I'll pull in that refactor as well.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 3 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:15]: Eh. I'm just going to pull in SemVer usage this PR.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 7 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:155]: Ah. Ok you meant `mkdir`. Yeah you're absolutely right. Gotta fix that.

File: eng/common/scripts/release-docs-metadata.ps1 (hunk 8 above)
Comment (scbedd) [reply eng/common/scripts/release-docs-metadata.ps1]: Refactor is pushed for a branch per readme. Gotta test the params in a test release though.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+- pwsh: |
+    git clone ${{ TargetDocRepository }} ${{ parameters.WorkingDirectory }}/repo
+    
+    $sourceBranch = "${{ parameters.TargetDocBranch }}"
+  
+    try {
+      Push-Location {{ parameters.WorkingDirectory }}/repo
+    
+      # 0 if exists, 1 if not
+      git show-ref --verify --quiet refs/heads/$sourceBranch
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:25]: Given you have a fresh clone a few steps above there should not be any existing branch other then master in the local clone. There might be a remote branch with the same name but what you are doing here isn't going to do what I think you want it to.

Comment (weshaggard) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:25]: Also if there does end up being a remote branch with the same name the forced reset at https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/git-branch-push.ps1#L108 will fix it up to rebase put the changes on top of that branch.

Do you expect that the branch will already exist in the normal case? If not then we should be good here but if so then we can make some changes to the git-branch-push script to look for a remote branch of the same name and base the checkout off of that if it exists.

File: eng/common/scripts/git-branch-push.ps1 (hunk 1 above)
Comment (scbedd) [reply eng/common/scripts/git-branch-push.ps1]: Alrighty. This will be configured in each release.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 10 above)
Comment (scbedd) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml]: Ok. In a normal case we expect `smoke-test` to exist. always. Hard error if not.

**Comment by scbedd** [comment]
Have this thing working down to [here](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=382227&view=logs&s=920b375f-847c-504d-8cee-9a1d03ea4bae&j=ceb84d66-6af3-5d95-9875-cd812f26b59e)

Wherein it's failing due to unshared history. It's definitely detecting changes in the correct file, need to mirror the operation locally and figure out why this is occurring.

And I'm a bit confused as to why the PR isn't submitting appropriately. It's picking up the changes for certain!

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 11]
```diff
@@ -6,33 +6,23 @@ parameters:
...
-      if (-not $sourceBranchExists) {
-        # there is an existing smoke-test branch we can base off of
-        Write-Host "git checkout $sourceBranch"
-        git checkout $sourceBranch
-      }
-
-      Write-Host "git checkout -b {{ parameters.PRBranchName }}"
-      git checkout -b ${{ parameters.PRBranchName }}
+      Write-Host "git checkout smoke-test"
+      git checkout smoke-test
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:25]: If there is ever a case where this branch doesn't exist in the repo this step will fail. Is that why you have the try/finally? Perhaps you have access to make this as a protected branch in the docs repo to help ensure it always exists.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 12]
```diff
@@ -43,16 +33,21 @@ steps:
   displayName: 'Apply Documentation Updates From Artifact'
   inputs:
     targetType: filePath
-    filePath: ${{ parameters.ScriptDirectory }}/release-docs-metadata.ps1
-    arguments: -ArtifactLocation ${{parameters.ArtifactLocation}} -Repository ${{parameters.PackageRepository}} -ReleaseSHA ${{parameters.ReleaseSha}} -RepoId ${{parameters.RepoId}} -WorkDirectory '${{parameters.WorkingDirectory}}' -DocRepoLocation "${{parameters.WorkingDirectory}}/repo" -Language "${{parameters.Language}}" -DocRepoContentLocation "docs-ref-services/"
+    filePath: ${{ parameters.ScriptDirectory }}/update-docs-metadata.ps1
+    arguments: -ArtifactLocation ${{parameters.ArtifactLocation}} -Repository ${{parameters.PackageRepository}} -ReleaseSHA ${{parameters.ReleaseSha}} -RepoId ${{parameters.RepoId}} -WorkDirectory '${{parameters.WorkingDirectory}}' -DocRepoLocation "${{parameters.WorkingDirectory}}/repo" -Language "${{parameters.Language}}" -DocRepoContentLocation ${{ parameters.DocRepoDestinationPath }}
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:37]: Could you turn this to multiple lines to help with approving the readability?

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 13]
```diff
@@ -43,16 +33,21 @@ steps:
...
     GH_TOKEN: $(azuresdk-github-pat)
 
 - template: /eng/common/pipelines/templates/steps/create-pull-request.yml
   parameters:
-    RepoName: ${{ parameters.TargetDocRepo }}
-    PRBranchName: increment-package-version-${{ parameters.ArtifactName }}-$(Build.BuildId)
+    RepoName: ${{ parameters.TargetDocRepoName }}
+    RepoOwner: ${{ parameters.TargetDocRepoOwner }}
+    PRBranchName: ${{ parameters.PRBranchName }}
     CommitMsg: "Increment package version after release of ${{ parameters.ArtifactName }}"
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:48]: I assume you want to update the commit message as well.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 14]
```diff
@@ -6,33 +6,23 @@ parameters:
   RepoId: 'not-specified'
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:6]: Looks like there a are a number of unused parameters.

File: eng/common/pipelines/templates/steps/create-pull-request.yml [hunk 15]
```diff
@@ -16,8 +16,8 @@ parameters:
 steps:
 
 - pwsh: |
-    echo "git add ."
-    git add .
+    echo "git add -A ."
+    git add -A .
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/create-pull-request.yml:20]: This is a good fix as we were missing deletions. I think this can be change to simply `git add -A` without the `.`. https://stackoverflow.com/questions/572549/difference-between-git-add-a-and-git-add/572660#572660 has a great demonstration about how this works.

File: eng/common/scripts/update-docs-metadata.ps1 [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+  $ReleaseSHA, # the SHA for the artifacts. DevOps: $(Release.Artifacts.<artifactAlias>.SourceVersion) or $(Build.SourceVersion)
+  $RepoId, # full repo id. EG azure/azure-sdk-for-net  DevOps: $(Build.Repository.Id). Used as a part of VerifyPackages
+  $Repository, # EG: "Maven", "PyPI", "NPM"
+
+  # arguments necessary to power the docs release
+  $DocRepoLocation, # the location on disk where we have cloned the documentation repository
+  $Language, # EG: js, java, dotnet. Used in language for the embedded readme.
+  $DocRepoContentLocation = "docs-ref-services/" # within the doc repo, where does our readme go?
+)
+
```
Comment (weshaggard) [inline eng/common/scripts/update-docs-metadata.ps1:16]: To help with debugging I suggest adding a line to output the command similar to what I added to other scripts see https://github.com/Azure/azure-sdk-tools/commit/e157cfa8c83cf55e65561b50d59f7cd6de5700e9.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 13 above)
Comment (scbedd) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:48]: On my local it _is_. 

image

Looking.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 12 above)
Comment (scbedd) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:37]: Ah. Only reason I hadn't was because I thought the `arguments` needed to be a single line, not an array. Thought this was a yml

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 14 above)
Comment (scbedd) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:6]: Just one. Fixed.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 11 above)
Comment (scbedd) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:25]: I intentionally designed it this way. `smoke-test` will always exist. I haven't taken the step to make it a protected branch as sometimes temporarily deletion is the easiest way to clean up the mountains of commited YML, but it will _always_ exist.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 12 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:36]: yml supports 2 types of line splitting folded (`>`) and literal (`|`) see https://yaml-multiline.info/. So I think you can change this to something like:

```
arguments: >
  -ArtifactLocation ${{parameters.ArtifactLocation}}
  -Repository ${{parameters.PackageRepository}}
  -ReleaseSHA ${{parameters.ReleaseSha}}
  -RepoId ${{parameters.RepoId}}
  -WorkDirectory '${{parameters.WorkingDirectory}}'
  -DocRepoLocation "${{parameters.WorkingDirectory}}/repo"
  -Language "${{parameters.Language}}"
  -DocRepoContentLocation ${{ parameters.DocRepoDestinationPath }}
```

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+  ignoreLASTEXITCODE: false
+
+- task: PowerShell@2
+  displayName: 'Apply Documentation Updates From Artifact'
+  inputs:
+    targetType: filePath
+    filePath: ${{ parameters.ScriptDirectory }}/update-docs-metadata.ps1
+    arguments: -ArtifactLocation ${{parameters.ArtifactLocation}} -Repository ${{parameters.PackageRepository}} -ReleaseSHA ${{parameters.ReleaseSha}} -RepoId ${{parameters.RepoId}} -WorkDirectory '${{parameters.WorkingDirectory}}' -DocRepoLocation "${{parameters.WorkingDirectory}}/repo" -Language "${{parameters.Language}}" -DocRepoContentLocation ${{ parameters.DocRepoDestinationPath }}
+    pwsh: true
+  timeoutInMinutes: 5
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:38]: Why explicitly set a 5 min timeout?

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 18]
```diff
@@ -0,0 +1,471 @@
+. (Join-Path $PSScriptRoot SemVer.ps1)
+
+$VERSION_REGEX = "(?<major>\d+)(\.(?<minor>\d+))?(\.(?<patch>\d+))?((?<pre>[^0-9][^\s]+))?"
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:3]: Is this still needed when you pull in SemVer? If we are missing some versioning helpers maybe better to add them to SemVer.

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 19]
```diff
@@ -0,0 +1,471 @@
...
+    Invoke-WebRequest-WithHandling -url $url -body $body -headers $headers -method "Post"
+  }
+}
+
+function Invoke-WebRequest-WithHandling($url, $method, $body = $null, $headers = $null) {
+  $attempts = 1
+
+  while ($attempts -le 3) {
+    try {
+      return Invoke-RestMethod -Method $method -Uri $url -Body $body -Headers $headers
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:46]: Is there a reason not to use the MaxRetryCount option natively on Invoke-RestMethod?

**Comment by scbedd** [comment]
This PR just keeps getting longer. :/

I'll make a pass for more fixes I guess.

[Successful release](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=383457&view=results)

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 19 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:46]: I knew about MaxRetryCount. At the time I'm pretty certain we wanted explicit output for the reset information.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 17 above)
Comment (scbedd) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:38]: Good point. No guarantee some of these clones will run this fast anyway.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 18 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1]: 100%. (ha)

I'm now completely utilizing the SemVer.ps1 code. Notice that I've made a SLIGHT change to that to make the regex static so that people outside the class can leverage the resource.

**Comment by scbedd** [comment]
@weshaggard Thanks. I've made the appropriate adjustments for SemVer as well. You've already signed off, but want to make certain that you see that change.

File: eng/common/scripts/artifact-metadata-parsing.ps1 (hunk 7 above)
Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1]: @weshaggard this is fixed. now. All `mkdir` have been converted over to `New-Item` calls with `-Force` attached to ensure that missing parents are created.

File: eng/common/scripts/artifact-metadata-parsing.ps1 [hunk 20]
```diff
@@ -0,0 +1,472 @@
+. (Join-Path $PSScriptRoot SemVer.ps1)
+
+$SDIST_PACKAGE_REGEX = "^(?<package>.*)\-(?<versionstring>$([AzureEngSemanticVersion]::SEMVER_REGEX.Substring(1)))"
```
Comment (weshaggard) [inline eng/common/scripts/artifact-metadata-parsing.ps1:3]: The substring(1) is a little odd and might break in the future. It might be better to remove the `^` and `$` from the REGEX especially if we plan to use it in ways like this.

Comment (scbedd) [reply eng/common/scripts/artifact-metadata-parsing.ps1:3]: I'll push one additional commit.

File: eng/common/scripts/SemVer.ps1 [hunk 21]
```diff
@@ -19,7 +19,7 @@ class AzureEngSemanticVersion {
     [bool] $IsPrerelease
     [string] $RawVersion
     # Regex inspired but simplifie from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
-    static [string] $SEMVER_REGEX = "^(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)(?:-?(?<prelabel>[a-zA-Z-]*)(?:\.?(?<prenumber>0|[1-9]\d*)))?$"
+    static [string] $SEMVER_REGEX = "(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)(?:-?(?<prelabel>[a-zA-Z-]*)(?:\.?(?<prenumber>0|[1-9]\d*)))?"
```
Comment (weshaggard) [inline eng/common/scripts/SemVer.ps1:22]: Since you removed the begin/end here you might want to add them in the ParseVersionString method below just so we can be sure we are still only parsing a version.

File: eng/common/scripts/SemVer.ps1 [hunk 22]
```diff
@@ -24,7 +24,7 @@ class AzureEngSemanticVersion {
     static [AzureEngSemanticVersion] ParseVersionString([string] $versionString)
     {
         try {
-            return [AzureEngSemanticVersion]::new($versionString)
+            return [AzureEngSemanticVersion]::new("^$versionString$")
```
Comment (weshaggard) [inline eng/common/scripts/SemVer.ps1:27]: I was meaning to add it around the regex not the version string.

File: eng/common/scripts/SemVer.ps1 [hunk 23]
```diff
@@ -31,10 +32,7 @@ class AzureEngSemanticVersion {
     }
     
     AzureEngSemanticVersion([string] $versionString){
-        # Regex inspired but simplifie from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
-        $SEMVER_REGEX = "^(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)(?:-?(?<prelabel>[a-zA-Z-]*)(?:\.?(?<prenumber>0|[1-9]\d*)))?$"
-        
-        if ($versionString -match $SEMVER_REGEX) {
+        if ($versionString -match "^$([AzureEngSemanticVersion]::SEMVER_REGEX)$") {
```
Comment (weshaggard) [inline eng/common/scripts/SemVer.ps1:35]: Out of curiosity did you get a chance to validate this? I'm not sure if PS will like the unescaped `$` at the end.
